### PR TITLE
[CHANGE][NRPTI#1211] Add intermediate code for act name

### PR DESCRIPTION
**Pull request by sanjaytkbabu**
## Pull Request Standards

- [x] The title of the PR is accurate
- [x] The title includes the type of change [`HOTFIX`, `FEATURE`, `etc`]  
- [x] The PR title includes the ticket number in format of `[NRPTI-###]`
- [x] Documentation is updated to reflect change

# Description
In order to sure that records always display the most current act name without developer intervention, an intermediate value that is mapped to the most up-to-date value is attached to the records instead of the actual value.

This PR includes the following proposed change(s):

- a migration that updates all records with the actName "Energy Resource Activities Act" to instead have their actName = "ACT_103"
- updated the front end to check if a record has a code or the actual name. If a code, translate it to the actual value for displaying.
- updated the front end so that search filters can can appropriately catch results with either the actual value or the appropriate code.
- updated the 'add record' functionality so that all records added using it are written with the code instead of the actual value.
- updated the data structure that the front end uses to access all legislation values from {actName: [regulations]} TO {actCode : { actName, [regulations] } } and refactored related functions to handle this new structure.

Context for PR review :

There are many files changed in this PR (44?!) , BUT most of them are identical changes. 

   - All the add-edit components have been updated to write the intermediate code rather than the actual act name
   - All the record detail components have been updated to pass the service that can translate the intermediate code down to the utility that builds all the legislation details (including act name) that are displayed when viewing a record.
   -  angular/projects/common/src/app/utils/utils.ts is where that service gets used to translate a code into a name
   - angular/projects/admin-nrpti/src/app/records/records-resolver.ts is where the actCode is added alongside the actName when building a search query. This makes it so searches for an act name return records with the intermediate code in place of the actual act name
   - the data-services are what need to be refactored to handle the restructured code-to-legislation data

Also, while the commit history seems daunting, it includes commits from the PR that this work is based on. As that PR has been merged in, the changes don't show up in files changed. Only the last 8-10 commits are for this ticket, starting from 2 weeks ago

### CODE REVIEW

File: angular/projects/admin-nrpti/src/app/records/administrative-penalties/administrative-penalty-add-edit/administrative-penalty-add-edit.component.ts [hunk 1]
```diff
@@ -542,6 +543,11 @@ export class AdministrativePenaltyAddEditComponent implements OnInit, OnDestroy
     // tslint:disable-next-line:max-line-length
     this.myForm.get('legislations').dirty &&
       (administrativePenalty['legislation'] = this.parseLegislationsFormGroups());
+    // swapping legislation with actCode
+    const actTitle = administrativePenalty['legislation'][0]['act'];
```
Comment (fergmac) [inline angular/projects/admin-nrpti/src/app/records/administrative-penalties/administrative-penalty-add-edit/administrative-penalty-add-edit.component.ts:547]: Do we need to check for existence of `administrativePenalty['legislation'` before indexing?

File: angular/projects/admin-nrpti/src/app/records/administrative-sanctions/administrative-sanction-add-edit/administrative-sanction-add-edit.component.ts [hunk 2]
```diff
@@ -457,7 +457,11 @@ export class AdministrativeSanctionAddEditComponent implements OnInit, OnDestroy
     // tslint:disable-next-line:max-line-length
     this.myForm.get('legislations').dirty &&
       (administrativeSanction['legislation'] = this.parseLegislationsFormGroups());
-
+    // swapping legislation with actCode
+    const actTitle = administrativeSanction['legislation'][0]['act'];
```
Comment (fergmac) [inline angular/projects/admin-nrpti/src/app/records/administrative-sanctions/administrative-sanction-add-edit/administrative-sanction-add-edit.component.ts:461]: Need to check before indexing here?

Comment (fergmac) [reply angular/projects/admin-nrpti/src/app/records/administrative-sanctions/administrative-sanction-add-edit/administrative-sanction-add-edit.component.ts:461]: In all further instances below as well?

Comment (fergmac) [reply angular/projects/admin-nrpti/src/app/records/administrative-sanctions/administrative-sanction-add-edit/administrative-sanction-add-edit.component.ts:461]: Would it make sense for this to be a util function? Easier to test and update.

File: angular/projects/admin-nrpti/src/app/records/records-resolver.ts [hunk 3]
```diff
@@ -55,6 +56,15 @@ export class RecordsResolver implements Resolve<Observable<object>> {
 
     if (params.act) {
       or['legislation.act'] = params.act;
+      // adds the actCode associated with each act to the query obj
```
Comment (fergmac) [inline angular/projects/admin-nrpti/src/app/records/records-resolver.ts:59]: We aren't repeating this code like above but could also be a util. Again easier to test and update.

File: angular/projects/common/src/app/utils/utils.ts [hunk 4]
```diff
@@ -8,15 +12,32 @@ export class Utils {
...
+    const ACT_CODE_BEGINNING = 'ACT_';
     if (!obj) {
       return '';
     }
 
     const legistrationStrings = [];
-
     if (obj.act) {
-      legistrationStrings.push(obj.act);
+      if (obj.act.substring(0, ACT_CODE_BEGINNING.length) === ACT_CODE_BEGINNING) {
```
Comment (fergmac) [inline angular/projects/common/src/app/utils/utils.ts:23]: Again this could be an act util. The comment could be reflected in the name of the function.

**Review by fergmac — APPROVED** [review]

File: angular/projects/admin-nrpti/src/app/records/administrative-sanctions/administrative-sanction-add-edit/administrative-sanction-add-edit.component.ts (hunk 2 above)
Comment (acatchpole) [reply angular/projects/admin-nrpti/src/app/records/administrative-sanctions/administrative-sanction-add-edit/administrative-sanction-add-edit.component.ts:461]: created a function in record-utils to take this object and swap values in place. also included error handling. refactored all these add-edit components to use this new function

File: angular/projects/admin-nrpti/src/app/records/administrative-penalties/administrative-penalty-add-edit/administrative-penalty-add-edit.component.ts (hunk 1 above)
Comment (acatchpole) [reply angular/projects/admin-nrpti/src/app/records/administrative-penalties/administrative-penalty-add-edit/administrative-penalty-add-edit.component.ts:547]: see below

File: angular/projects/admin-nrpti/src/app/records/records-resolver.ts (hunk 3 above)
Comment (acatchpole) [reply angular/projects/admin-nrpti/src/app/records/records-resolver.ts]: created a function in record-utils to do this and refactored.

File: angular/projects/common/src/app/utils/utils.ts (hunk 4 above)
Comment (acatchpole) [reply angular/projects/common/src/app/utils/utils.ts]: added an isActCode method and refactored

**Review by fergmac — APPROVED** [review]

**Pull request merged**